ree/reeEstimateError.cc

```diff
@@ -0,0 +1,458 @@
+    if(marker_list[i])
+      ess_dofs[ess_dof_counter++] = i;
+    else
+      uness_dofs[uness_dof_counter++] = i;
+  }
+}
+
+/**
+ * Inputs: Matrix A, Vector X, Vector B, essential dofs, other dofs.
+ * Output: reduced matrix A, reduced rhs B.
+ */
+static void eliminateDBCs(
+    mth::Matrix<double> const &A,
+    mth::Vector<double> const &X,
+    mth::Vector<double> const &B,
+    apf::NewArray<int>  const &ess_dofs,
+    apf::NewArray<int>  const &uness_dofs,
+    mth::Matrix<double> &Anew,
+    mth::Vector<double> &Bnew)
+{
+  int num_ess_dofs = ess_dofs.size();
+  int num_uness_dofs = uness_dofs.size();
+
+  // 1. Remove rows of A corresponding to
+	// ess_dofs by copying into Anew
+  Anew.resize(num_uness_dofs, num_uness_dofs);
+	for(int rr = 0; rr < num_uness_dofs; rr++) {
+    int i = uness_dofs[rr];
+    for(int cc = 0; cc < num_uness_dofs; cc++) {
+      int j = uness_dofs[cc];
+      Anew(rr,cc) = A(i,j);
+    }
+  }
+
+	// 2. Assemble new B
+	Bnew.resize(num_uness_dofs);
+  for(int i = 0; i < num_uness_dofs; i++) {
+    Bnew(i) = B(uness_dofs[i]);
+  }
+
+  if (num_ess_dofs > 0) {
+	  // 3. Subtract from Bnew: (Bnew -= Ae*Xe)
+    mth::Matrix<double> Ae(num_uness_dofs, num_ess_dofs);
+    for(int rr = 0; rr < num_uness_dofs; rr++) {
+      int i = uness_dofs[rr];
+      for(int cc = 0; cc < num_ess_dofs; cc++) {
+        int j = ess_dofs[cc];
+        Ae(rr,cc) = A(i,j);
+      }
+    }
+
+    mth::Vector<double> Xe(num_ess_dofs);
+    for(int i = 0; i < num_ess_dofs; i++) {
+      Xe(i) = X(ess_dofs[i]);
+    }
+
+    mth::Vector<double> temp;
+    mth::multiply(Ae, Xe, temp);
+    Bnew -= temp;
+  }
+}
+
+static double computeL2Error(apf::Mesh* mesh, apf::MeshEntity* e,
+	apf::Field* f, mth::Vector<double> const error_dofs)
+{
+	double error = 0.0;
+
+  apf::FieldShape* fs = f->getShape();
+  int type = mesh->getType(e);
+	PCU_ALWAYS_ASSERT(type == apf::Mesh::TET);
+  int nd = apf::countElementNodes(fs, type);
+  int dim = apf::getDimension(mesh, e);
+  double w;
+
+  apf::MeshElement* me = apf::createMeshElement(mesh, e);
+  apf::Element* el = apf::createElement(f, me);
+  int order = 2 * fs->getOrder();
+  int np = apf::countIntPoints(me, order);
+
+  apf::NewArray<apf::Vector3> vectorshape(nd);
+
+  apf::Vector3 p;
+  for (int i = 0; i < np; i++) {
+    apf::getIntPoint(me, order, i, p);
+    double weight = apf::getIntWeight(me, order, i);
+    apf::Matrix3x3 J;
+    apf::getJacobian(me, p, J);
+    double jdet = apf::getJacobianDeterminant(J, dim);
+    w = weight * jdet;
+
+    apf::getVectorShapeValues(el, p, vectorshape);
+    mth::Matrix<double> vectorShape (nd, dim);
+    for (int j = 0; j < nd; j++)
+      for (int k = 0; k < dim; k++)
+        vectorShape(j,k) = vectorshape[j][k];
+
+    mth::Matrix<double> vectorShapeT (dim, nd);
+    mth::transpose(vectorShape, vectorShapeT);
+
+		mth::Vector<double> err_func;
+		mth::multiply(vectorShapeT, error_dofs, err_func);
+
+		error += w * (err_func * err_func);
+	}
```

Comment:
@Samiullah-Malik 

There seems to be a problem with indentation here.

Reply:
Resolved.